cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/input/BasicPartitionedFileSetInputContext.java

```diff
@@ -42,34 +49,74 @@
 
-  private final Path inputPath;
-  private final String mappingString;
+  private final Map<String, PartitionKey> pathToPartitionMapping;
 
-  private PartitionKey partitionKey;
+  private final boolean isCombineInputFormat;
+  private final Configuration conf;
+
+  private final List<Path> inputPaths;
+  private List<PartitionKey> partitionKeys;
+
+  // for caching in case of CombineFileInputFormat
+  private String currentInputfileName;
+  private PartitionKey currentPartitionKey;
 
   BasicPartitionedFileSetInputContext(MultiInputTaggedSplit multiInputTaggedSplit) {
     super(multiInputTaggedSplit.getName());
 
     InputSplit inputSplit = multiInputTaggedSplit.getInputSplit();
-    if (!(inputSplit instanceof FileSplit)) {
-      throw new IllegalArgumentException(String.format("Expected a '%s', but got '%s'.",
-                                                       FileSplit.class.getName(), inputSplit.getClass().getName()));
+    inputPaths = new ArrayList<>();
+    if (inputSplit instanceof FileSplit) {
+      isCombineInputFormat = false;
+      inputPaths.add(((FileSplit) inputSplit).getPath());
+    } else if (inputSplit instanceof CombineFileSplit) {
+      isCombineInputFormat = true;
+      Collections.addAll(inputPaths, ((CombineFileSplit) inputSplit).getPaths());
+    } else {
+      throw new IllegalArgumentException(String.format("Expected either a '%s' or a '%s', but got '%s'.",
+                                                       FileSplit.class.getName(), CombineFileSplit.class.getName(),
+                                                       inputSplit.getClass().getName()));
     }
-    this.inputPath = ((FileSplit) inputSplit).getPath();
-    this.mappingString = multiInputTaggedSplit.getConf().get(PartitionedFileSetDataset.PATH_TO_PARTITIONING_MAPPING);
+
+    this.conf = multiInputTaggedSplit.getConf();
+    String mappingString = conf.get(PartitionedFileSetDataset.PATH_TO_PARTITIONING_MAPPING);
+    this.pathToPartitionMapping =
+      GSON.fromJson(Objects.requireNonNull(mappingString), STRING_PARTITION_KEY_MAP_TYPE);
   }
 
   @Override
   public PartitionKey getInputPartitionKey() {
-    if (partitionKey == null) {
-      partitionKey = getPartitionKey(inputPath.toUri());
+    if (isCombineInputFormat) {
+      // org.apache.hadoop.mapreduce.lib.input.CombineFileRecordReader sets this in its initNextRecordReader method
+      String inputFileName = conf.get(MRJobConfig.MAP_INPUT_FILE);
+      if (!inputFileName.equals(currentInputfileName)) {
+        Preconditions.checkNotNull(inputFileName);
+        currentPartitionKey = getPartitionKey(URI.create(inputFileName));
+        currentInputfileName = inputFileName;
+      }
+      return currentPartitionKey;
+    } else {
+      List<PartitionKey> inputPartitionKeys = getInputPartitionKeys();
+      if (inputPartitionKeys.size() != 1) {
+        throw new IllegalArgumentException(String.format("Expected only a single PartitionKey, but found: %s",
```

Comment:
I think this is an IllegalState, not an IllegalArgument. Can this actually happen? Also, if it's 0, the "only" is misleading... perhaps remove that word from the message?